Adds sentient spiders to midround events

benbot16 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
This is a partial port of tgstation/tgstation#54451.

This PR replaces the AI-controlled spider event with an event that spawns sentient spiders. It also reworks how spiders function a bit.

The event itself works by spawning 2 midwife spiders, using the same spawn mechanics as xenomorphs. These midwives immediately get the ability to lay a cluster of eggs, giving them some initial numbers, since midwives are squishy.

Egg clusters themselves have also been changed. They now don't hatch into spiderlings, instead allowing ghosts to interact with them after a minute to become a spider of their choice. If a spider eats a human, the spiders available will be one more powerful spider (tarantula/midwife/viper) and 3 less powerful spiders (base spider/nurse/hunter). If they eat something that's not a human, it'll just be 4 less-powerful spiders.

There's also been a few changes to spiders themselves. Spiders have been sped up, but they now slow down as they take damage. Nurses and midwives have also been given the ability to heal other spiders. AI-controlled spiders are also generally slower than player-controlled ones. Spiders also can't speak common anymore, instead speaking buzzwords, but they can still understand common. 

Additionally, spiders now have a new type of venom. This venom does less damage, but will deal stamina damage over time based on how much is in someone's system. It also has a chance to paralyze, and paralyzes more rapidly if there's purgatives in someone's system. 

New values for each spider type:
Giant spider: 100 health, 25 damage, 1 speed, -0.1 on web, 0 poison
Hunter: 80 health, 20 damage, 0 speed, 5 poison
Viper: 50 health, 1 damage, 0 speed, 12 poison (venom)
Nurse: 40 health, 10 damage, 1 speed, 0.5 on web, 3 poison, faster webbing
Midwife: 80 health, 10 damage, 2 speed, 1 on web, 3 poison, fastest webbing
Tarantula: 300 health, 35 damage, 5 speed off web, 0 speed on web, fast webbing

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->
The spider infestation event is really lame right now, and this would improve it. This also improves spider AI a bit, so they'll be a bit more interesting to deal with when exploring or dealing with spiderlings created by other things.

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>
The event:

https://user-images.githubusercontent.com/9423435/194977423-546a1b6c-a778-4f48-862b-f9d76021c795.mp4

Eggs:

https://user-images.githubusercontent.com/9423435/194977468-344adcd1-e664-453f-9f6c-39352756f035.mp4

Spider healing:

https://user-images.githubusercontent.com/9423435/194977695-9e5f6ad7-c803-49cc-8339-c315275b02bd.mp4



</details>

## Changelog
:cl: IndieanaJones, Rukofamicom, Benbot16
add: The Spider infestation midround event now spawns sentient spiders. These fall under the "spider" midround preference.
add: Nurse spiders can now heal other spiders.
tweak: Webs can now be stacked, up to a limit of 3.
tweak: Reworks how spider eggs spawn new spiders.
tweak: Spiders now take damage slowdown below 75% health.
tweak: Spider venom will now cause stamina damage, and paralyzes in large doses or if purged with chemicals.
tweak: Spiders now speak buzzwords. They can still understand common.
tweak: Tweaks spider AI.
balance: Tweaks several spider stats, mostly decreasing health and increasing speed.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

benbot16 committed:
Adds a sentient spider event

This adds a sentient version of the Spider Infestation event. It also updates directives logic to allow for the playable spiders created by this event to set directives and lay eggs.

benbot16 committed:
Updates sentient spiders - Draft

benbot16 committed:
More changes to how spiders work

benbot16 committed:
More spider fixes

benbot16 committed:
Changes to code wording/comments

benbot16 committed:
File change for health

benbot16 committed:
Tweaks some values and adds directives

benbot16 committed:
Some fixes/tweaks

benbot16 committed:
a delta_time fix

benbot16 committed:
Adds spawners menu handling and fixes the AI not working

benbot16 committed:
Some spider stat changes

benbot16 committed:
comment grammar

*scream

benbot16 committed:
Poison adjustment

benbot16 committed:
Adds damage slowdown to spiders

benbot16 committed:
webfix

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -629,3 +629,46 @@
+	antag_flag_override = ROLE_ALIEN
+	midround_ruleset_style = MIDROUND_RULESET_STYLE_HEAVY
+	required_type = /mob/dead/observer
+	enemy_roles = list("Security Officer", "Detective", "Head of Security", "Captain")
```
this is not consistent with our other enemy lists and doesn't use job name defines

benbot16 replied:
Done

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -629,3 +629,46 @@
+			if(temp_vent_parent.other_atmosmch.len > 20)
+				vents += temp_vent // Makes sure the pipeline is large enough
+	if(!vents.len)
+		return FALSE
```
make this consistent with xenos midround, there's a log message about execute failure. also use length() instead of len

benbot16 replied:
Done

benbot16 committed:
Updates the spider midround ruleset

benbot16 committed:
Consistency

benbot16 committed:
Adds a special spider venom for spiders

Code courtesy of Rukofamicom

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -629,3 +629,47 @@
+
+/datum/dynamic_ruleset/midround/from_ghosts/spiders
+	name = "Spider Infestation"
+	antag_flag = "Spider"
```
make a ROLE_SPIDER

benbot16 replied:
Made one. I'm not entirely sure that I set it up properly, though.

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -629,3 +629,47 @@
+	midround_ruleset_style = MIDROUND_RULESET_STYLE_HEAVY
+	required_type = /mob/dead/observer
+	enemy_roles = list(JOB_NAME_SECURITYOFFICER, JOB_NAME_DETECTIVE, JOB_NAME_WARDEN, JOB_NAME_HEADOFSECURITY, JOB_NAME_CAPTAIN)
+	required_enemies = list(2,2,1,1,1,0,0,0,0,0)
```
change to 2,2,1,1,1,1,1,0,0,0

benbot16 replied:
Done

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -629,3 +629,47 @@
+	required_enemies = list(2,2,1,1,1,0,0,0,0,0)
+	required_candidates = 2
+	weight = 5
+	cost = 10
```
lower cost to 8

benbot16 replied:
Done

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -629,3 +629,47 @@
+	enemy_roles = list(JOB_NAME_SECURITYOFFICER, JOB_NAME_DETECTIVE, JOB_NAME_WARDEN, JOB_NAME_HEADOFSECURITY, JOB_NAME_CAPTAIN)
+	required_enemies = list(2,2,1,1,1,0,0,0,0,0)
+	required_candidates = 2
+	weight = 5
```
do 4

benbot16 replied:
Done

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -629,3 +629,47 @@
+	required_candidates = 2
+	weight = 5
+	cost = 10
+	repeatable = TRUE
```
add minimum_players = 27

benbot16 replied:
Done

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -629,3 +629,47 @@
+			var/datum/pipeline/temp_vent_parent = temp_vent.parents[1]
+			if(!temp_vent_parent)
+				continue // No parent vent
+			if(temp_vent_parent.other_atmosmch.len > 20)
```
len >:( use length()

benbot16 replied:
Done

benbot16 committed:
Adds ROLE_SPIDER, some .len replacements, dynamic midround value tweaks

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -27,29 +27,34 @@
-	melee_damage = 15
+	initial_language_holder = /datum/language_holder/spider // Speaks buzzwords, understands buzzwords and common
+	maxHealth = 100
+	health = 100
```
```suggestion
	maxHealth = 75
	health = 75
```

100 is pretty high as far as HP goes, putting them on par with the average human player. Given just how easily spiders can make more spiders, the average spider should not be on par with a human 1v1 unless you want them to regularly snowball out of control.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -27,29 +27,34 @@
+	maxHealth = 100
+	health = 100
+	obj_damage = 25
+	melee_damage = 25
```
```suggestion
	melee_damage = 15
```

25 is a also a LOT of damage, even when compared to the power of an above average crewmember. 15 is a much better baseline for a "standard" spider, especially if you follow through with making them venomous too.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -27,29 +27,34 @@
+	health = 100
+	obj_damage = 25
+	melee_damage = 25
+	poison_per_bite = 0
```
```suggestion
	poison_per_bite = 4
```

Spiders make a more unique threat if the majority utilize and rely on venom rather than bursty brute force.

benbot16 replied:
I intentionally had the "brown" spiders be stronger and not have venom, so that they're differentiated from the hunter/viper spiders. That way it encourages working together a bit more, with the hunters being able to paralyze but not quite having the stopping power of the slower and tankier brown spiders. I'll probably have to tweak damage for both of these more to get to that point, though.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -70,32 +76,68 @@
 	melee_damage = 10
 	poison_per_bite = 3
+	speed = 1 // A bit faster than midwives
 	var/atom/movable/cocoon_target
+	var/mob/living/simple_animal/hostile/poison/giant_spider/heal_target
 	var/fed = 0
+	var/enriched_fed = 0
 	var/obj/effect/proc_holder/wrap/wrap
 	var/datum/action/innate/spider/lay_eggs/lay_eggs
 	var/datum/action/innate/spider/set_directive/set_directive
 	var/static/list/consumed_mobs = list() //the tags of mobs that have been consumed by nurse spiders to lay eggs
 	gold_core_spawnable = NO_SPAWN
```
Everything related to reproduction and feeding should be shifted down and be exclusive to the Broodmother (/midwife)
New Broodmothers should be blocked from spawning via gold slimes and spiderlings, rather than nurses.

benbot16 replied:
I left it on nurses to make make spiders be more decentralized. Having to deal with nurses too removes the possibility of just nuking the couple of broodmothers that exist using a flamer and being done with it.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -318,6 +282,123 @@
 
+// Midwives are upgraded nurses. They can web quickly and are stronger than regular nurses, but they're a bit slower.
+/mob/living/simple_animal/hostile/poison/giant_spider/nurse/midwife
+	name = "midwife"
```
```suggestion
	name = "broodmother"
```

Superior name that better communicates the spider's role, especially to newer players. It's also a name that many players already utilize despite not being the actual name.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -318,6 +282,123 @@
+	icon_state = "midwife"
+	icon_living = "midwife"
+	icon_dead = "midwife_dead"
+	maxHealth = 80
+	health = 80
```
```suggestion
	maxHealth = 200
	health = 200
```

Following the logic of making it impossible to produce more broodmothers, they will need a lot more HP given their lacking combat capability.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -318,6 +282,123 @@
+	maxHealth = 80
+	health = 80
+	melee_damage = 20
+	poison_per_bite = 5
```
```suggestion
	maxHealth = 60
	health = 60
	melee_damage = 15
	poison_per_bite = 6
```

Following the description and changes to the base spider, this follows the intention of a slightly squishier but more venomous variant than the standard giant spider.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -318,6 +282,123 @@
+	maxHealth = 50
+	health = 50
+	melee_damage = 1
+	poison_per_bite = 12
```
```suggestion
	maxHealth = 35
	health = 35
	melee_damage = 5
	poison_per_bite = 12
```

Lethal in as little as one bite with patience and thought, they deserve to be squishier too. 5 melee_damage means they actually finish a victim paralyzed by venom in a reasonable timeframe while still needing to rely almost entirely on venom in order to kill.

benbot16 replied:
I left these guys with standard venom, instead of spider venom. They still have one damage because of that. 
I'm not entirely sure how good of an idea it is, but I wanted these guys to still be pretty lethal.

Ideally I'd still like to keep them at this stat level, since their health means that getting hit will put some pretty decent damage slowdown on them, and giving them less health will make it even easier for them to just die to atmos changes without being able to move.

Rukofamicom commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -318,6 +282,123 @@
+	icon_dead = "tarantula_dead"
+	maxHealth = 300
+	health = 300
+	melee_damage = 35
```
```suggestion
	maxHealth = 160
	health = 160
	melee_damage = 25
	poison_per_bite = 0
```

As the primary defenders they deserve to have more HP, but 300 is absurdly high for a mob that is intended by design to protect and move in close proximity with others capable of healing it.

benbot16 committed:
Tweaks some spider stats

benbot16 commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -70,32 +76,68 @@
 	health = 40
 	melee_damage = 10
 	poison_per_bite = 3
+	speed = 1 // A bit faster than midwives
```
They already web at a quarter of normal speed.

Rukofamicom replied:
I see where that's defined now, my mistake. Can't seem to close these, I may not even be able to - this is my first time leaving any sort of review.

benbot16 committed:
Adds a spider antag datum + teams; Does some spider balance

benbot16 committed:
Master tracking for spiders

benbot16 committed:
Adds spider antag HUD handling

benbot16 committed:
HUD update

benbot16 committed:
Adds spider IFF

benbot16 committed:
Reverts a NPC change

This is probably too unperformant, and thinking about it we probably don't need all that much handling for AI spidermobs.

benbot16 committed:
Some IFF fixes

benbot16 committed:
Fixes some dynamic stuff

benbot16 committed:
Adjusts spider venom again

Went a bit to heavy on it last time, gonna give it a bit of time at this level to get more info on it.

benbot16 committed:
Greetings and directives refactor

benbot16 committed:
Gives more spiders web speed bonuses, buffs webs/tile

benbot16 committed:
Bigger command text

benbot16 committed:
Bit more clarification to the greet

benbot16 committed:
Adds an IPC check

benbot16 committed:
Fixes spider end-rounds

benbot16 committed:
Nerfs poison a bit and adds something for burning eggclusters

benbot16 committed:
Nerfs web armor vs. brute and adds a retaliation check to cocooning.

benbot16 committed:
fix for attackedby and a spider spawn rate nerf

benbot16 committed:
Grow time tweak

benbot16 committed:
cycle check

benbot16 committed:
Poison tweak

benbot16 committed:
Metab change

benbot16 committed:
Tweaks enriched spawns logic and a name fix.

PestoVerde322 reviewed (dismissed):
Icon seems fine, waiting for code approval.

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -642,3 +642,59 @@
+	required_enemies = list(2,2,1,1,1,1,1,0,0,0)
+	required_candidates = 2
+	weight = 4
+	cost = 8
```
Dynamic 2023: Raise cost to 11

benbot16 replied:
Done

itsmeow commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -642,3 +642,59 @@
+	enemy_roles = list(JOB_NAME_SECURITYOFFICER, JOB_NAME_DETECTIVE, JOB_NAME_WARDEN, JOB_NAME_HEADOFSECURITY, JOB_NAME_CAPTAIN)
+	required_enemies = list(2,2,1,1,1,1,1,0,0,0)
+	required_candidates = 2
+	weight = 4
```
Dynamic 2023: Weight 3

benbot16 replied:
Done

benbot16 committed:
Weight changes

itsmeow reviewed (dismissed):
As is, I am happy enough with spiders and their function. They have good QoL and function as an antagonist, however their balance still leaves a bit to desire. I think this is better worked out with another PR than repeated testmerging as they still fundamentally work and we desperately need more midround variety. I am going to approve this, although this does not mean I think they are perfect balance wise. I would like to see more of how strategy plays out on the spider side, or what 'meta' evolves so we can work on making it more engaging and unique.

itsmeow commented on code/modules/events/spider_infestation.dm:
```diff
@@ -1,40 +1,63 @@
-/datum/round_event/spider_infestation
-	announceWhen	= 400
+/datum/round_event/ghost_role/spider_infestation
+	role_name = "spider"
```
this should be ROLE_SPIDER

benbot16 replied:
Done

benbot16 committed:
role name
